brooklyn-server/core/src/main/java/org/apache/brooklyn/core/entity/AbstractApplication.java

```diff
@@ -220,13 +221,37 @@ public void stop() {
                 //but that could be handled by the impl at management
                 //(keeping recently unmanaged things)  
                 //  however unmanaging must be done last, _after_ we stop children and set attributes 
+                Location machineProvisioningLocation = Iterables.getFirst(getLocations(), null);
                 getEntityManager().unmanage(this);
+                if (machineProvisioningLocation != null) {
+                    unmanageProvisionerIfNotUsed(machineProvisioningLocation);
+                }
             }
         }
 
         logApplicationLifecycle("Stopped");
     }
 
+    //TODO Think of general approach for expunging location provisioners.
+    // I left a code which could be a possible general approach. However it needs tests.
+    protected boolean unmanageProvisionerIfNotUsed(/*MachineProvisioningLocation<MachineLocation>*/Location machineProvisioningLocation) {
```

Comment:
Can you write a simple test that demonstrates that the location was being left as managed (i.e. a failing test before this change, and passes after this change).